File: doc-examples/example-java/src/test/java/example/BookMethods.java
package example;

import io.micronaut.el.ELMethodContributor;
import io.micronaut.el.ELMethodRegistry;

public final class BookMethods implements ELMethodContributor { // <1>

    @Override
    public void contribute(ELMethodRegistry registry) {
        registry
            .method(Book.class, "title", String.class, Book::getTitle) // <2>
            .method(Book.class, "discounted", double.class, Double.class, Book::discounted) // <3>
            .method(Book.class, "label", String.class, String.class, Double.class,
                (Book book, String currency, Double percent) -> currency + book.discounted(percent)) // <4>
            .method(Book.class, "tagged", String.class, new Class<?>[]{String.class, String[].class}, true,
                (context, base, arguments) -> ((Book) base).getTitle()
                    + " " + String.join((String) arguments[0], (String[]) arguments[1])) // <5>
            .method(Book.class, "summarised", String.class, Summary.class,
                (Book book, Summary summary) -> summary.of(book)) // <6>
            .staticMethod(Math.class, "abs", long.class, long.class, Math::abs) // <7>
            .constructor(Book.class, new Class<?>[]{String.class, String.class, double.class}, false,
                (context, base, arguments) ->
                    new Book((String) arguments[0], (String) arguments[1], (Double) arguments[2])) // <8>
            .function("fmt", "shout", TextFunctions.class, "shout", String.class, String.class,
                TextFunctions::shout) // <9>
            .functionalInterface(Summary.class,
                (context, lambda) -> book -> String.valueOf(lambda.invoke(context, book))); // <10>
    }

    @Override
    public int getOrder() { // <11>
        return -100;
    }
}

File: doc-examples/example-python/src/test/python/example/BookMethods.py
import java
from java.lang import Double, Long, Math, String
from micronaut.el import ELMethodContributor, ELMethodRegistry

from example.Book import Book
from example.Summary import Summary
from example.TextFunctions import TextFunctions

# TODO(python): the primitive and array class literals have no import form
StringArray = java.type("java.lang.String[]")
double = java.type("double")
long = java.type("long")


class BookMethods(ELMethodContributor):  # <1>

    def contribute(self, registry: ELMethodRegistry) -> None:
        (registry
            .method(Book, "title", String, lambda book: book.title)  # <2>
            .method(Book, "discounted", double, Double, lambda book, percent: book.discounted(percent))  # <3>
            .method(Book, "label", String, String, Double,
                    lambda book, currency, percent: currency + str(book.discounted(percent)))  # <4>
            .method(Book, "tagged", String, [String, StringArray], True,
                    lambda context, base, arguments: base.title + " " + arguments[0].join(arguments[1]))  # <5>
            .method(Book, "summarised", String, Summary, lambda book, summary: summary.of(book))  # <6>
            .staticMethod(Math, "abs", long, Long, lambda value: Math.abs(value))  # <7>
            .constructor(Book, [String, String, double], False,
                         lambda context, base, arguments: Book(arguments[0], arguments[1], arguments[2]))  # <8>
            .function("fmt", "shout", TextFunctions, "shout", String, String,
                      lambda text: TextFunctions.shout(text))  # <9>
            .functionalInterface(Summary,
                                 lambda context, lambda_: LambdaSummary(context, lambda_)))  # <10>

    def getOrder(self) -> int:  # <11>
        return -100


class LambdaSummary(Summary):
    """The Summary a lambda expression becomes: invokes the lambda with the book."""

    def __init__(self, context, lambda_):
        self.context = context
        self.lambda_ = lambda_

    def of(self, book: Book) -> str:
        return str(self.lambda_.invoke(self.context, book))
